app/controllers/concerns/course/assessment/live_feedback/message_concern.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+module Course::Assessment::LiveFeedback::MessageConcern
+  extend ActiveSupport::Concern
+
+  def handle_save_user_message
+    @thread = Course::Assessment::LiveFeedback::Thread.where(codaveri_thread_id: @thread_id).first
+
+    new_message = Course::Assessment::LiveFeedback::Message.create!({
+      thread_id: @thread.id,
+      is_error: false,
+      content: @message,
+      creator_id: current_user.id,
+      created_at: Time.zone.now,
+      option_id: nil # TODO: change after we handle the suggestion passing to BE
+    })
+
+    if answer_updated_from_last_snapshot_files(new_message)
+      create_new_snapshot_files_and_associate_with_new_message(new_message)
+    else
+      associate_new_message_with_existing_files(new_message)
+    end
+  end
+
+  def answer_updated_from_last_snapshot_files(new_message)
+    last_message = @thread.messages.where.not(id: new_message.id).max_by(&:id)
+    return true unless last_message
+
+    file_ids = Course::Assessment::LiveFeedback::MessageFile.where(message_id: last_message.id).pluck(:file_id)
+
+    file_hash = Course::Assessment::LiveFeedback::File.where(id: file_ids).to_h { |file| [file.filename, file.content] }
+
+    @answer.actable.files.each do |file|
+      return true unless file_hash[file.filename] == file.content
+    end
+
+    false
+  end
+
+  def associate_new_message_with_existing_files(new_message)
+    last_message = @thread.messages.where.not(id: new_message.id).max_by(&:id)
+    file_ids = Course::Assessment::LiveFeedback::MessageFile.where(message_id: last_message.id).pluck(:file_id)
+
+    file_ids.each do |file_id|
+      Course::Assessment::LiveFeedback::MessageFile.create!({
+        message_id: new_message.id,
+        file_id: file_id
+      })
+    end
+  end
+
+  def create_new_snapshot_files_and_associate_with_new_message(new_message)
+    @answer.actable.files.each do |programming_file|
+      file = Course::Assessment::LiveFeedback::File.create!({
+        filename: programming_file.filename,
+        content: programming_file.content
+      })
+
+      Course::Assessment::LiveFeedback::MessageFile.create!({
+        message_id: new_message.id,
+        file_id: file.id
+      })
+    end
```

Comment:
These type of multiple creates shouldn't be done in a loop, and should be done as a bulk create, or in a transaction.